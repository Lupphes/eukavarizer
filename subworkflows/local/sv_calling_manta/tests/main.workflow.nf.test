/*
 * SV_CALLING_MANTA Workflow Tests
 *
 * Manta is a structural variant caller optimized for Illumina paired-end and mate-pair reads.
 * Key features:
 * - Designed specifically for short-read Illumina sequencing data
 * - Requires paired-end reads (filters out single-end data)
 * - Optimized for reads shorter than the long_read_threshold (default 300bp)
 * - Produces three VCF outputs: small variants, candidate SVs, and diploid SVs
 * Platform support: Illumina paired-end reads only
 * Read length filtering: Filters out reads >= long_read_threshold
 */

nextflow_workflow {

    name "Test Workflow SV_CALLING_MANTA"
    script "../main.nf"
    workflow "SV_CALLING_MANTA"
    config "subworkflows/local/sv_calling_manta/tests/nextflow.config"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "sv_calling_manta"

    /*
     * Test scenario 1: Standard Illumina paired-end BAM input
     * This test validates Manta's basic SV calling with appropriate Illumina paired-end reads.
     * Expected behavior: Manta should process the BAM and generate three types of VCF outputs:
     * - small_vcf: Small variants and indels
     * - candidate_vcf: Candidate structural variants
     * - diploid_vcf: High-confidence diploid SVs
     * Each VCF includes compressed (vcfgz), tabix index (tbi), and CSI index (csi) files.
     */
    test("Illumina paired-end BAM input - stub") {
        options '-stub'

        when {
            params {
                long_read_threshold = 1000
            }
            workflow {
                """
                input[0] = channel.of([
                    [id: 'test', single_end: false, platform: 'illumina', median_bp: 150, sample: 'sample1', patient: 'patient1'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)])
                input[2] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.small_vcf != null },
                { assert workflow.out.small_vcfgz != null },
                { assert workflow.out.small_tbi != null },
                { assert workflow.out.candidate_vcf != null },
                { assert workflow.out.candidate_vcfgz != null },
                { assert workflow.out.candidate_tbi != null },
                { assert workflow.out.diploid_vcf != null },
                { assert workflow.out.diploid_vcfgz != null },
                { assert workflow.out.diploid_tbi != null },
                { assert workflow.out.small_csi != null },
                { assert workflow.out.candidate_csi != null },
                { assert workflow.out.diploid_csi != null },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    /*
     * Test scenario 2: Single-end reads are filtered out
     * This test validates Manta's platform filtering logic - it should skip single-end data.
     * Expected behavior: The workflow should succeed but produce no output, as Manta requires
     * paired-end reads. Single-end reads (single_end: true) should be filtered before calling Manta.
     */
    test("Single-end reads filtered out - stub") {
        options '-stub'

        when {
            params {
                long_read_threshold = 1000
            }
            workflow {
                """
                // Single-end sample that should be filtered out (Manta requires paired-end)
                input[0] = channel.of([
                    [id: 'test_single', single_end: true, platform: 'illumina', median_bp: 150, sample: 'sample1', patient: 'patient1'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)])
                input[2] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success }
            )
        }
    }

    /*
     * Test scenario 3: Long reads are filtered out
     * This test validates Manta's read length filtering - it should skip long-read data.
     * Expected behavior: The workflow should succeed but produce no output, as Manta is optimized
     * for short reads. Reads with median_bp >= long_read_threshold (5000bp >= 1000bp) should be
     * filtered before calling Manta.
     */
    test("Long reads filtered out - stub") {
        options '-stub'

        when {
            params {
                long_read_threshold = 1000
            }
            workflow {
                """
                // ONT long-read sample (5000bp) that should be filtered out (threshold is 1000bp)
                input[0] = channel.of([
                    [id: 'test_ont', single_end: false, platform: 'ont', median_bp: 5000, sample: 'sample1', patient: 'patient1'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)])
                input[2] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success }
            )
        }
    }

    /*
     * Test scenario 4: Multiple samples processed in batch
     * This test validates Manta's ability to process multiple Illumina paired-end samples in parallel.
     * Expected behavior: Manta should successfully call SVs for each sample independently,
     * generating separate VCF outputs for each sample (sample1 and sample2), including
     * small variants, candidate SVs, and diploid SVs.
     */
    test("Multiple samples - stub") {
        options '-stub'

        when {
            params {
                long_read_threshold = 1000
            }
            workflow {
                """
                // Two Illumina paired-end samples from the same patient
                input[0] = channel.of(
                    [
                        [id: 'sample1', single_end: false, platform: 'illumina', median_bp: 150, sample: 'sample1', patient: 'patient1'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                    ],
                    [
                        [id: 'sample2', single_end: false, platform: 'illumina', median_bp: 150, sample: 'sample2', patient: 'patient1'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                    ]
                )
                input[1] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)])
                input[2] = channel.value([[id: 'genome'], file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.small_vcf != null },
                { assert workflow.out.candidate_vcf != null },
                { assert workflow.out.diploid_vcf != null },
            { assert workflow.out.small_csi != null },
            { assert workflow.out.candidate_csi != null },
            { assert workflow.out.diploid_csi != null }
            )
        }
    }
}
