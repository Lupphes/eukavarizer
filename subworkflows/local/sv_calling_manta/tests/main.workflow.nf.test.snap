{
    "Illumina paired-end BAM input - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "12": [
                    "versions.yml:md5,284d95f693c1028bd6077738aea041f0",
                    "versions.yml:md5,39d9730982a6860490a00b55d7deae76",
                    "versions.yml:md5,58621d88d18a28d6a37d14560defeb6e",
                    "versions.yml:md5,5b4c8d66fe7a6730ac69de13d10b44df",
                    "versions.yml:md5,66c0904c0a0ec6b4634f0b7c2f8909bb",
                    "versions.yml:md5,91c7046f4ac92ce08d0e22be4c740c01",
                    "versions.yml:md5,99eec772b6511235ef2d4b7174605235"
                ],
                "2": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "candidate_csi": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "candidate_tbi": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "candidate_vcf": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "candidate_vcfgz": [
                    [
                        {
                            "id": "test-manta-candidate-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "diploid_csi": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "diploid_tbi": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "diploid_vcf": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "diploid_vcfgz": [
                    [
                        {
                            "id": "test-manta-diploid-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "small_csi": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.csi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "small_tbi": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "small_vcf": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "small_vcfgz": [
                    [
                        {
                            "id": "test-manta-small-svync",
                            "single_end": false,
                            "platform": "illumina",
                            "median_bp": 150,
                            "sample": "sample1",
                            "patient": "patient1"
                        },
                        "re-sample1-illumina-false-manta.vcf.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,284d95f693c1028bd6077738aea041f0",
                    "versions.yml:md5,39d9730982a6860490a00b55d7deae76",
                    "versions.yml:md5,58621d88d18a28d6a37d14560defeb6e",
                    "versions.yml:md5,5b4c8d66fe7a6730ac69de13d10b44df",
                    "versions.yml:md5,66c0904c0a0ec6b4634f0b7c2f8909bb",
                    "versions.yml:md5,91c7046f4ac92ce08d0e22be4c740c01",
                    "versions.yml:md5,99eec772b6511235ef2d4b7174605235"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-23T02:38:56.35815357"
    }
}